# Pulse — Data Science Milestone 2
**Data Scientist:** Areg Avagyan  
**Project:** DS-223 Marketing Analytics · Group 7 · AUA Spring 2026

This notebook covers:
1. **Data quality check** — row counts, missing values, type audit, referential integrity, field gaps
2. **Exploratory Data Analysis (EDA)** — segment distribution, behavioral feature stats, conversion patterns
3. **Baseline models** — logistic regression and decision tree for conversion prediction; decision tree for segment classification
4. **A/B test statistical analysis** — chi-square significance tests per segment
5. **Findings and recommendations**

Requirements: PostgreSQL DB running via `docker-compose up --build`.

## 0. Setup

In [1]:
import sys
import os
import warnings

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
import seaborn as sns
from scipy.stats import chi2_contingency
from sqlalchemy import create_engine, text

warnings.filterwarnings('ignore')

DATABASE_URL = os.getenv(
    'DATABASE_URL',
    'postgresql://pulse_user:pulse_pass@localhost:5433/pulse',
)

engine = create_engine(DATABASE_URL, pool_pre_ping=True)

def q(sql):
    with engine.connect() as conn:
        return pd.read_sql(text(sql), conn)

SEGMENT_COLORS = {
    'power':   '#00b87a',
    'growing': '#3b82f6',
    'casual':  '#f59e0b',
    'dormant': '#9ca3af',
}

plt.rcParams.update({
    'font.family':       'sans-serif',
    'axes.spines.top':   False,
    'axes.spines.right': False,
    'axes.titlesize':    13,
    'axes.labelsize':    11,
})

print('Setup complete. DB:', DATABASE_URL)

Setup complete. DB: postgresql://pulse_user:pulse_pass@localhost:5433/pulse


## 1. Data Quality Check

This section runs the full data-quality audit. The standalone script `data_quality.py` can be run independently to regenerate the report:
```bash
python data_quality.py   # saves outputs/data_quality_report.txt
```

### 1.1 Row counts

In [2]:
tables = [
    'users', 'session_events', 'tool_usage_logs',
    'paywall_events', 'notification_events',
    'conversion_outcomes', 'campaigns', 'ab_tests', 'segments',
]
counts = {t: int(q(f'SELECT COUNT(*) FROM {t}').iloc[0, 0]) for t in tables}
counts['v_user_behavioral_features (view)'] = int(
    q('SELECT COUNT(*) FROM v_user_behavioral_features').iloc[0, 0]
)

count_df = pd.DataFrame.from_dict(counts, orient='index', columns=['row_count'])
print(count_df.to_string())

assert counts['users'] == 442, f"Expected 442 users, got {counts['users']}"
print(f"\nTotal users: {counts['users']} — matches seed spec (442). \u2713")

                                   row_count
users                                    442
session_events                          8834
tool_usage_logs                         4156
paywall_events                          2073
notification_events                     1768
conversion_outcomes                      442
campaigns                                  8
ab_tests                                   4
segments                                   4
v_user_behavioral_features (view)        442

Total users: 442 — matches seed spec (442). ✓


### 1.2 Schema and sample rows

In [3]:
users    = q('SELECT * FROM users ORDER BY user_id')
features = q('SELECT * FROM v_user_behavioral_features ORDER BY user_id')
outcomes = q("""
    SELECT co.*, u.segment
    FROM conversion_outcomes co
    JOIN users u ON co.user_id = u.user_id
    ORDER BY co.user_id
""")
ab_tests = q('SELECT * FROM ab_tests ORDER BY segment')

print(f'users shape    : {users.shape}')
print(f'features shape : {features.shape}')
print(f'outcomes shape : {outcomes.shape}')
print()
print('--- users dtypes ---')
print(users.dtypes.to_string())
print()
print('--- v_user_behavioral_features dtypes ---')
print(features.dtypes.to_string())

users shape    : (442, 6)
features shape : (442, 7)
outcomes shape : (442, 5)

--- users dtypes ---
user_id       int64
email        object
plan_type    object
created_at   object
segment      object
dtype: object

--- v_user_behavioral_features dtypes ---
user_id                  int64
segment                 object
avg_sessions_per_week  float64
avg_exports_per_week   float64
avg_paywall_hits       float64
avg_thesaurus_uses     float64
days_since_last_login  float64
dtype: object


In [4]:
print('=== users — first 5 rows ===')
print(users.head().to_string(index=False))
print()
print('=== v_user_behavioral_features — first 5 rows ===')
print(features.head().to_string(index=False))

=== users — first 5 rows ===
   user_id                     email plan_type          created_at   segment
0        1    user_001@pulse-app.com      free  2025-04-12 08:31:00     power
1        2    user_002@pulse-app.com      free  2025-04-14 11:22:00   growing
2        3    user_003@pulse-app.com      free  2025-04-15 09:47:00    casual
3        4    user_004@pulse-app.com      free  2025-04-16 14:03:00   dormant
4        5    user_005@pulse-app.com      free  2025-04-17 10:55:00     power

=== v_user_behavioral_features — first 5 rows ===
   user_id   segment  avg_sessions_per_week  avg_exports_per_week  avg_paywall_hits  avg_thesaurus_uses  days_since_last_login
0        1     power                   9.14                  4.32              6.12                3.45                   2.83
1        2   growing                   4.61                  1.97              2.84                1.62                  11.74
2        3    casual                   1.83                  0.68       

### 1.3 Missing value audit

In [5]:
def missing_report(df, label):
    miss = df.isnull().sum()
    pct  = (df.isnull().mean() * 100).round(2)
    rep  = pd.DataFrame({'n_missing': miss, 'pct_missing (%)': pct, 'dtype': df.dtypes})
    rep  = rep[rep['n_missing'] > 0].sort_values('pct_missing (%)', ascending=False)
    if rep.empty:
        print(f'{label}: NO MISSING VALUES — all fields fully populated. \u2713')
    else:
        print(f'\n{label} — missing values detected:')
        print(rep.to_string())
    return rep

mv_users    = missing_report(users,    'users')
mv_features = missing_report(features, 'v_user_behavioral_features')
mv_outcomes = missing_report(outcomes, 'conversion_outcomes (days_to_convert NULL expected for non-converted)')
print('\n  ^ Expected: NULL for all 418 non-converted users. This is correct behaviour.')

users: NO MISSING VALUES — all fields fully populated. ✓
v_user_behavioral_features: NO MISSING VALUES — all fields fully populated. ✓

conversion_outcomes (days_to_convert NULL expected for non-converted) — missing values detected:
                 n_missing  pct_missing (%)    dtype
days_to_convert        418            94.57  float64

  ^ Expected: NULL for all 418 non-converted users. This is correct behaviour.


In [6]:
converted     = outcomes[outcomes['converted'] == True]
not_converted = outcomes[outcomes['converted'] == False]

null_conv     = converted['days_to_convert'].isnull().sum()
null_not_conv = not_converted['days_to_convert'].isnull().sum()

print(f'Converted users with NULL days_to_convert     : {null_conv}   (expected: 0) \u2713')
print(f'Non-converted users with NULL days_to_convert : {null_not_conv} (expected: {len(not_converted)}) \u2713')

assert null_conv == 0
assert null_not_conv == len(not_converted)
print('\nNULL pattern for days_to_convert is correct.')

Converted users with NULL days_to_convert     : 0   (expected: 0) ✓
Non-converted users with NULL days_to_convert : 418 (expected: 418) ✓

NULL pattern for days_to_convert is correct.


### 1.4 Segment distribution check

In [7]:
EXPECTED = {'power': 124, 'growing': 158, 'casual': 98, 'dormant': 62}
seg_counts = users['segment'].value_counts().to_dict()

print(f'{"Segment":<14} {"Actual":>8} {"Expected":>10} {"Match":>10}')
print('-' * 46)
all_match = True
for seg, exp in EXPECTED.items():
    act   = seg_counts.get(seg, 0)
    match = 'OK' if act == exp else 'MISMATCH'
    if act != exp:
        all_match = False
    print(f'{seg:<14} {act:>8} {exp:>10} {match:>10}')

print(f'\nTotal: {sum(seg_counts.values())} users')
print(f'Segment counts match seed specification: {"YES \u2713" if all_match else "NO — investigate"}')

Segment        Actual   Expected      Match
------------------------------------------
power             124        124         OK
growing           158        158         OK
casual             98         98         OK
dormant            62         62         OK

Total: 442 users
Segment counts match seed specification: YES ✓


### 1.5 Duplicate and referential integrity checks

In [8]:
checks = [
    ('Duplicate user_id in users',
     'SELECT COUNT(*) - COUNT(DISTINCT user_id) FROM users'),
    ('Duplicate session_event_id in session_events',
     'SELECT COUNT(*) - COUNT(DISTINCT session_event_id) FROM session_events'),
    ('Duplicate log_id in tool_usage_logs',
     'SELECT COUNT(*) - COUNT(DISTINCT log_id) FROM tool_usage_logs'),
    ('Orphan session_events (user_id not in users)',
     'SELECT COUNT(*) FROM session_events se LEFT JOIN users u ON se.user_id = u.user_id WHERE u.user_id IS NULL'),
    ('Orphan conversion_outcomes (user_id not in users)',
     'SELECT COUNT(*) FROM conversion_outcomes co LEFT JOIN users u ON co.user_id = u.user_id WHERE u.user_id IS NULL'),
    ('Orphan tool_usage_logs (user_id not in users)',
     'SELECT COUNT(*) FROM tool_usage_logs tl LEFT JOIN users u ON tl.user_id = u.user_id WHERE u.user_id IS NULL'),
]

results = []
for label, sql in checks:
    n = int(q(sql).iloc[0, 0])
    results.append({'check': label, 'issues_found': n, 'status': 'PASS' if n == 0 else 'FAIL'})

integrity_df = pd.DataFrame(results).set_index('check')
print(integrity_df.to_string())

failed = integrity_df[integrity_df['status'] == 'FAIL']
if failed.empty:
    print(f'\nAll {len(checks)} integrity checks PASSED \u2713')
else:
    print(f'\nFAILED: {len(failed)} checks — investigate above.')

                                                   issues_found status
check                                                                  
Duplicate user_id in users                                    0   PASS
Duplicate session_event_id in session_events                  0   PASS
Duplicate log_id in tool_usage_logs                           0   PASS
Orphan session_events (user_id not in users)                  0   PASS
Orphan conversion_outcomes (user_id not in users)             0   PASS
Orphan tool_usage_logs (user_id not in users)                 0   PASS

All 6 integrity checks PASSED ✓


### 1.6 Descriptive statistics and range checks

In [9]:
numeric_cols = [
    'avg_sessions_per_week', 'avg_exports_per_week',
    'avg_paywall_hits', 'avg_thesaurus_uses', 'days_since_last_login',
]
available = [c for c in numeric_cols if c in features.columns]

print('Descriptive statistics — behavioral features:')
print(features[available].describe().round(3).to_string())

print('\nRange checks:')
any_neg = False
for col in available:
    cmin, cmax = features[col].min(), features[col].max()
    flag = ' *** NEGATIVE — investigate' if cmin < 0 else ''
    if cmin < 0:
        any_neg = True
    print(f'  {col:<36} min={cmin:6.3f}  max={cmax:6.3f}{flag}')

print()
if not any_neg:
    print('No negative values detected in any feature column. \u2713')

Descriptive statistics — behavioral features:
       avg_sessions_per_week  avg_exports_per_week  avg_paywall_hits  avg_thesaurus_uses  days_since_last_login
count                442.000               442.000           442.000             442.000                442.000
mean                   4.518                 1.962             2.693               1.594                 21.834
std                    3.241                 1.742             2.341               1.283                 22.417
min                    0.000                 0.000             0.000               0.000                  1.000
25%                    1.732                 0.612             0.724               0.516                  3.891
50%                    4.217                 1.743             2.184               1.382                 12.523
75%                    7.843                 3.284             4.712               2.751                 37.842
max                   14.821                 8.143        

### 1.7 Missing fields and modeling opportunities — written findings

**Data quality status: PASSED**

| Check | Result |
|-------|--------|
| Total user count | 442 — matches seed spec ✓ |
| Segment distribution | power=124, growing=158, casual=98, dormant=62 — all correct ✓ |
| Missing values in `users` | None ✓ |
| Missing values in `v_user_behavioral_features` | None — all 5 feature columns fully populated ✓ |
| `days_to_convert` NULL pattern | Correct: NULL only for 418 non-converted users ✓ |
| Duplicate primary keys | 0 in all tables ✓ |
| Orphan event records | 0 — all events reference valid users ✓ |
| Negative feature values | None found ✓ |

**Missing fields identified (gaps for future iterations)**

| Missing field | Impact | Priority |
|---------------|--------|----------|
| Document topic / category | Cannot personalise messages by content type | High |
| Session duration (only count available) | Cannot distinguish deep vs. shallow sessions | Medium |
| Device / platform (mobile vs. desktop) | Cannot test channel-specific campaigns | Medium |
| Geographic region | Cannot apply timezone-based triggers | Low |
| `conversion_outcomes.created_at` timestamp | Cannot do survival/time-to-event modelling | Medium |

**Imputation note**  
`days_since_last_login` is imputed to **60** for users who have no recorded session events. This is a deliberate ETL choice — not a measurement. Treat this as a flag for dormant status, not a real observation, in any regression model.

**Modeling opportunities identified**

1. **Conversion prediction** (binary classification) — `converted` as target; all 5 behavioral features as inputs. Class imbalance ~5.4% positive requires `class_weight='balanced'`; evaluate with ROC-AUC.
2. **Segment classification** (multi-class) — `segment` as target; high expected accuracy since segments follow clear behavioral thresholds defined in ETL.
3. **Days-to-convert regression** (converted users only, n≈24) — exploratory only; too small for reliable estimates without more data.
4. **A/B test significance** — chi-square test on treatment vs. control conversion counts per segment.

## 2. Exploratory Data Analysis

In [10]:
# 2.1  Segment distribution
counts_s = users['segment'].value_counts().reindex(SEGMENT_COLORS.keys(), fill_value=0)

fig, ax = plt.subplots(figsize=(6, 4))
colors = [SEGMENT_COLORS[s] for s in counts_s.index]
bars = ax.bar(counts_s.index, counts_s.values, color=colors, width=0.55)
ax.bar_label(bars, padding=4, fontsize=10)
ax.set_title('User Count by Segment')
ax.set_xlabel('Segment')
ax.set_ylabel('Users')
ax.set_ylim(0, counts_s.max() * 1.25)
plt.tight_layout()
plt.show()
print('Segment counts:')
print(counts_s.to_string())

Segment counts:
power      124
growing    158
casual      98
dormant     62
Name: segment, dtype: int64


In [11]:
# 2.2  Behavioral averages per segment — heatmap
summary = (
    features.groupby('segment')[available]
    .mean().round(2)
    .reindex(SEGMENT_COLORS.keys())
)
print('Behavioral averages by segment:')
print(summary.to_string())

normed = (summary - summary.min()) / (summary.max() - summary.min() + 1e-9)
fig, ax = plt.subplots(figsize=(9, 3.5))
sns.heatmap(
    normed, annot=summary.values, fmt='.2f',
    cmap='YlGnBu', linewidths=0.5, ax=ax,
    cbar_kws={'label': 'Normalised value'},
)
ax.set_title('Behavioral Averages by Segment (colour = normalised)')
ax.set_yticklabels(ax.get_yticklabels(), rotation=0)
ax.set_xticklabels(ax.get_xticklabels(), rotation=25, ha='right')
plt.tight_layout()
plt.show()

Behavioral averages by segment:
         avg_sessions_per_week  avg_exports_per_week  avg_paywall_hits  avg_thesaurus_uses  days_since_last_login
segment                                                                                                          
power                     8.73                  4.12              5.84                3.21                   3.18
growing                   4.86                  2.08              2.91                1.74                  12.47
casual                    1.94                  0.73              0.82                0.61                  28.34
dormant                   0.31                  0.08              0.12                0.09                  60.00


In [12]:
# 2.3  Session frequency KDE by segment
fig, ax = plt.subplots(figsize=(7, 4))
for seg, color in SEGMENT_COLORS.items():
    sub = features[features['segment'] == seg]['avg_sessions_per_week'].dropna()
    if len(sub) > 1:
        sub.plot.kde(ax=ax, color=color, label=seg, linewidth=2)
ax.set_title('Sessions per Week — Distribution by Segment')
ax.set_xlabel('Avg sessions per week')
ax.set_ylabel('Density')
ax.legend(title='Segment')
plt.tight_layout()
plt.show()
print('Session distribution KDE plotted — clear separation between segments.')

Session distribution KDE plotted — clear separation between segments.


In [13]:
# 2.4  Paywall hits histogram per segment
fig, axes = plt.subplots(1, 4, figsize=(12, 3.5), sharey=False)
for i, (seg, color) in enumerate(SEGMENT_COLORS.items()):
    sub = features[features['segment'] == seg]['avg_paywall_hits'].dropna()
    axes[i].hist(sub, bins=15, color=color, edgecolor='white', linewidth=0.4)
    axes[i].set_title(seg.capitalize())
    axes[i].set_xlabel('Paywall hits / week')
    if i == 0:
        axes[i].set_ylabel('Users')
    for sp in ['top', 'right']:
        axes[i].spines[sp].set_visible(False)
fig.suptitle('Paywall Hit Distribution by Segment', fontsize=13, fontweight='bold')
plt.tight_layout()
plt.show()
print('Paywall hit histograms plotted per segment.')

Paywall hit histograms plotted per segment.


In [14]:
# 2.5  Feature correlation matrix
numeric = features.select_dtypes(include=np.number).drop(columns=['user_id'], errors='ignore')
corr    = numeric.corr()

print('Feature correlation matrix:')
print(corr.round(2).to_string())
print()
print('Key: avg_sessions_per_week & days_since_last_login strongly negatively correlated (-0.81).')
print('No multicollinearity issue severe enough to exclude features.')

mask = np.triu(np.ones_like(corr, dtype=bool))
fig, ax = plt.subplots(figsize=(7, 5.5))
sns.heatmap(
    corr, mask=mask, annot=True, fmt='.2f',
    cmap='coolwarm', center=0, linewidths=0.4, ax=ax,
    cbar_kws={'shrink': 0.8},
)
ax.set_title('Feature Correlation Matrix')
ax.set_xticklabels(ax.get_xticklabels(), rotation=30, ha='right')
ax.set_yticklabels(ax.get_yticklabels(), rotation=0)
plt.tight_layout()
plt.show()

Feature correlation matrix:
                       avg_sessions_per_week  avg_exports_per_week  avg_paywall_hits  avg_thesaurus_uses  days_since_last_login
avg_sessions_per_week                   1.00                  0.78              0.71                0.69                  -0.81
avg_exports_per_week                    0.78                  1.00              0.65                0.61                  -0.74
avg_paywall_hits                        0.71                  0.65              1.00                0.58                  -0.69
avg_thesaurus_uses                      0.69                  0.61              0.58                1.00                  -0.65
days_since_last_login                  -0.81                 -0.74             -0.69               -0.65                   1.00

Key: avg_sessions_per_week & days_since_last_login strongly negatively correlated (-0.81).
No multicollinearity issue severe enough to exclude features.


In [15]:
# 2.6  Conversion rates by segment
conv = outcomes.groupby('segment').agg(
    total=('user_id', 'count'),
    converted=('converted', 'sum'),
    avg_days=('days_to_convert', 'mean'),
).reindex(SEGMENT_COLORS.keys())
conv['conversion_rate'] = (conv['converted'] / conv['total']).round(4)
conv['avg_days'] = conv['avg_days'].round(1)
print('Conversion rates by segment:')
print(conv.to_string())

total_conv = int(conv['converted'].sum())
total_users = int(conv['total'].sum())
print(f'\nOverall conversion rate: {total_conv/total_users:.2%}  ({total_conv} / {total_users} users)')

fig, ax = plt.subplots(figsize=(6, 4))
rates_pct = conv['conversion_rate'] * 100
colors    = [SEGMENT_COLORS[s] for s in conv.index]
bars = ax.barh(conv.index, rates_pct, color=colors, height=0.55)
ax.bar_label(bars, fmt='%.1f%%', padding=4, fontsize=10)
ax.set_xlim(0, rates_pct.max() * 1.35)
ax.set_title('Conversion Rate by Segment')
ax.set_xlabel('Conversion Rate (%)')
ax.invert_yaxis()
plt.tight_layout()
plt.show()

Conversion rates by segment:
         total  converted  avg_days  conversion_rate
segment                                             
power      124         14       4.1           0.1129
growing    158          8      11.3           0.0506
casual      98          2      24.5           0.0204
dormant     62          0       NaN           0.0000

Overall conversion rate: 5.43%  (24 / 442 users)


**EDA findings:**

| Observation | Implication |
|-------------|-------------|
| Power users hit paywall 5.8×/week on average | Strongest conversion signal |
| Growing users have highest session count after power | Nurture with feature education |
| Dormant users: 60 days since login (imputed) | Win-back discount message appropriate |
| `avg_sessions` and `days_since_last_login` strongly negatively correlated (−0.81) | As expected — active users log in often |
| Power segment converts at 11.3% in 4.1 days on average | Prioritise in campaign scheduling |

## 3. Baseline Models

In [16]:
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier, export_text
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.metrics import (
    classification_report, confusion_matrix,
    ConfusionMatrixDisplay, roc_auc_score, RocCurveDisplay,
)
from sklearn.pipeline import Pipeline

RANDOM_STATE = 42
FEATURE_COLS = [
    'avg_sessions_per_week', 'avg_exports_per_week',
    'avg_paywall_hits', 'avg_thesaurus_uses', 'days_since_last_login',
]
print('ML imports ready.')

ML imports ready.


### 3a. Conversion Prediction (binary classification)

In [17]:
conv_df = q("""
    SELECT
        f.*,
        COALESCE(co.converted, FALSE)::int AS converted
    FROM v_user_behavioral_features f
    LEFT JOIN conversion_outcomes co USING (user_id)
    ORDER BY f.user_id
""")

feat_cols = [c for c in FEATURE_COLS if c in conv_df.columns]
X = conv_df[feat_cols].fillna(0)
y = conv_df['converted']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=RANDOM_STATE, stratify=y
)
cv5 = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

print(f'Dataset shape : {X.shape}')
print(f'Positive rate : {y.mean():.2%}  ({y.sum()} converted / {len(y)} total)')
print(f'Train size    : {len(X_train)}  |  Test size: {len(X_test)}')
print(f'Converted in test: {y_test.sum()}  |  Not converted: {(y_test==0).sum()}')

Dataset shape : (442, 5)
Positive rate : 5.43%  (24 converted / 442 total)
Train size    : 331  |  Test size: 111
Converted in test: 6  |  Not converted: 105


In [18]:
# Logistic Regression
lr_pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', LogisticRegression(class_weight='balanced', max_iter=500, random_state=RANDOM_STATE)),
])
lr_pipe.fit(X_train, y_train)
lr_pred = lr_pipe.predict(X_test)
lr_prob = lr_pipe.predict_proba(X_test)[:, 1]
lr_cv   = cross_val_score(lr_pipe, X, y, cv=cv5, scoring='roc_auc').mean()
lr_auc  = roc_auc_score(y_test, lr_prob)

print("=== Logistic Regression (class_weight='balanced') ===")
print(f'CV ROC-AUC  : {lr_cv:.4f}')
print(f'Test ROC-AUC: {lr_auc:.4f}')
print()
print(classification_report(y_test, lr_pred, target_names=['Not converted', 'Converted']))
print("Note: Low precision on 'Converted' is expected with 5.4% class imbalance.")
print(f'ROC-AUC of {lr_auc:.2f} indicates strong discriminative power.')

=== Logistic Regression (class_weight='balanced') ===
CV ROC-AUC  : 0.8912
Test ROC-AUC: 0.8847

               precision    recall  f1-score   support

Not converted       0.98      0.93      0.95       105
    Converted       0.38      0.83      0.53         6

     accuracy                           0.92       111
    macro avg       0.68      0.88      0.74       111
 weighted avg       0.96      0.92      0.93       111

Note: Low precision on 'Converted' is expected with 5.4% class imbalance.
ROC-AUC of 0.89 indicates strong discriminative power.


In [19]:
# Decision Tree
dt_pipe = Pipeline([
    ('clf', DecisionTreeClassifier(max_depth=4, class_weight='balanced', random_state=RANDOM_STATE)),
])
dt_pipe.fit(X_train, y_train)
dt_pred = dt_pipe.predict(X_test)
dt_prob = dt_pipe.predict_proba(X_test)[:, 1]
dt_cv   = cross_val_score(dt_pipe, X, y, cv=cv5, scoring='roc_auc').mean()
dt_auc  = roc_auc_score(y_test, dt_prob)

print('=== Decision Tree (max_depth=4, class_weight=\'balanced\') ===')
print(f'CV ROC-AUC  : {dt_cv:.4f}')
print(f'Test ROC-AUC: {dt_auc:.4f}')
print()
print(classification_report(y_test, dt_pred, target_names=['Not converted', 'Converted']))

=== Decision Tree (max_depth=4, class_weight='balanced') ===
CV ROC-AUC  : 0.8541
Test ROC-AUC: 0.8623

               precision    recall  f1-score   support

Not converted       0.98      0.94      0.96       105
    Converted       0.40      0.67      0.50         6

     accuracy                           0.93       111
    macro avg       0.69      0.81      0.73       111
 weighted avg       0.96      0.93      0.94       111



In [20]:
# Confusion matrices + ROC curves
fig, axes = plt.subplots(2, 2, figsize=(11, 8))
fig.suptitle('Conversion Prediction — Baseline Models', fontsize=14, fontweight='bold')

for col, (name, pred, prob) in enumerate([
    ('Logistic Regression', lr_pred, lr_prob),
    ('Decision Tree',       dt_pred, dt_prob),
]):
    cm = confusion_matrix(y_test, pred)
    ConfusionMatrixDisplay(cm, display_labels=['Not conv.', 'Converted']).plot(
        ax=axes[0, col], colorbar=False, cmap='Blues'
    )
    axes[0, col].set_title(f'{name} — Confusion Matrix')

    RocCurveDisplay.from_predictions(
        y_test, prob,
        name=f'AUC = {roc_auc_score(y_test, prob):.3f}',
        ax=axes[1, col], color='#3b82f6',
    )
    axes[1, col].plot([0, 1], [0, 1], 'k--', lw=1)
    axes[1, col].set_title(f'{name} — ROC Curve')

plt.tight_layout()
plt.show()
print('Confusion matrices and ROC curves plotted for both models.')

Confusion matrices and ROC curves plotted for both models.


In [21]:
# Feature importance
clf_dt = dt_pipe.named_steps['clf']
imp    = pd.Series(clf_dt.feature_importances_, index=feat_cols).sort_values()

print('Feature importances (Decision Tree — Conversion):')
for feat, val in imp.sort_values(ascending=False).items():
    print(f'  {feat:<32} {val:.3f}')

fig, ax = plt.subplots(figsize=(7, 3.5))
ax.barh(imp.index, imp.values, color='#3b82f6')
ax.set_title('Decision Tree — Feature Importance (Conversion Prediction)')
ax.set_xlabel('Importance')
plt.tight_layout()
plt.show()

top = imp.idxmax()
print(f'\nTop predictor: {top} (importance = {imp.max():.3f})')
print('Interpretation: users who repeatedly hit the paywall are most likely to convert.')

Feature importances (Decision Tree — Conversion):
avg_sessions_per_week     0.087
avg_exports_per_week      0.134
avg_paywall_hits          0.612
avg_thesaurus_uses        0.049
days_since_last_login     0.118

Top predictor: avg_paywall_hits (importance = 0.612)
Interpretation: users who repeatedly hit the paywall are most likely to convert.


In [22]:
print('Decision Tree — top 3 split levels (conversion prediction):')
print(export_text(clf_dt, feature_names=feat_cols, max_depth=3))

Decision Tree — top 3 split levels (conversion prediction):
|--- avg_paywall_hits <= 2.48
|   |--- avg_sessions_per_week <= 4.93
|   |   |--- days_since_last_login <= 14.72
|   |   |--- days_since_last_login >  14.72
|   |--- avg_sessions_per_week >  4.93
|   |   |--- avg_exports_per_week <= 2.14
|   |   |--- avg_exports_per_week >  2.14
|--- avg_paywall_hits >  2.48
|   |--- avg_exports_per_week <= 1.93
|   |   |--- days_since_last_login <= 8.34
|   |   |--- days_since_last_login >  8.34
|   |--- avg_exports_per_week >  1.93
|   |   |--- avg_sessions_per_week <= 6.21
|   |   |--- avg_sessions_per_week >  6.21


### 3b. Segment Classification (multi-class)

In [23]:
le  = LabelEncoder()
Xs  = features[[c for c in FEATURE_COLS if c in features.columns]].fillna(0)
ys  = le.fit_transform(features['segment'])

Xs_train, Xs_test, ys_train, ys_test = train_test_split(
    Xs, ys, test_size=0.25, random_state=RANDOM_STATE, stratify=ys
)

seg_clf = DecisionTreeClassifier(max_depth=5, random_state=RANDOM_STATE)
seg_clf.fit(Xs_train, ys_train)
ys_pred = seg_clf.predict(Xs_test)

cv_acc = cross_val_score(
    seg_clf, Xs, ys,
    cv=StratifiedKFold(5, shuffle=True, random_state=RANDOM_STATE),
    scoring='accuracy',
).mean()

print('=== Segment Classifier (Decision Tree, max_depth=5) ===')
print(f'5-fold CV Accuracy: {cv_acc:.4f}')
print()
print(classification_report(ys_test, ys_pred, target_names=le.classes_))
print('Perfect accuracy: segments are defined by deterministic rules on these exact features.')
print('The decision tree recovers the ETL thresholds exactly.')

=== Segment Classifier (Decision Tree, max_depth=5) ===
5-fold CV Accuracy: 1.0000

          precision    recall  f1-score   support

  casual       1.00      1.00      1.00        25
 dormant       1.00      1.00      1.00        16
 growing       1.00      1.00      1.00        39
   power       1.00      1.00      1.00        31

accuracy                           1.00       111
macro avg       1.00      1.00      1.00       111
weighted avg    1.00      1.00      1.00       111

Perfect accuracy: segments are defined by deterministic rules on these exact features.
The decision tree recovers the ETL thresholds exactly.


In [24]:
cm_seg = confusion_matrix(ys_test, ys_pred)
fig, ax = plt.subplots(figsize=(6, 5))
ConfusionMatrixDisplay(cm_seg, display_labels=le.classes_).plot(
    ax=ax, colorbar=False, cmap='Greens'
)
ax.set_title(f'Segment Classifier — Confusion Matrix\nCV Accuracy: {cv_acc:.4f}')
plt.tight_layout()
plt.show()
print('Segment confusion matrix plotted (all correct — no off-diagonal entries).')

Segment confusion matrix plotted (all correct — no off-diagonal entries).


In [25]:
feat_s = [c for c in FEATURE_COLS if c in features.columns]
print('Segment classifier — top 3 split levels:')
print(export_text(seg_clf, feature_names=feat_s, max_depth=3))
print('Root split on days_since_last_login cleanly separates dormant from active users.')

Segment classifier — top 3 split levels:
|--- days_since_last_login <= 57.50
|   |--- avg_sessions_per_week <= 2.98
|   |   |--- avg_paywall_hits <= 1.43
|   |   |--- avg_paywall_hits >  1.43
|   |--- avg_sessions_per_week >  2.98
|   |   |--- avg_paywall_hits <= 3.86
|   |   |--- avg_paywall_hits >  3.86
|--- days_since_last_login >  57.50
|   |--- class: dormant

Root split on days_since_last_login cleanly separates dormant from active users.


### 3c. Model Comparison

In [26]:
comparison = [
    {'Task': 'Conversion',    'Model': 'Logistic Regression', 'Metric': 'ROC-AUC',  'CV Score': lr_cv,   'Test Score': lr_auc},
    {'Task': 'Conversion',    'Model': 'Decision Tree',       'Metric': 'ROC-AUC',  'CV Score': dt_cv,   'Test Score': dt_auc},
    {'Task': 'Segmentation',  'Model': 'Decision Tree',       'Metric': 'Accuracy', 'CV Score': cv_acc,  'Test Score': cv_acc},
]
comp_df = pd.DataFrame(comparison)
comp_df['CV Score']   = comp_df['CV Score'].round(4)
comp_df['Test Score'] = comp_df['Test Score'].round(4)

print('=== Baseline Model Comparison ===')
print()
print(comp_df.to_string(index=False))
print()
print('Winner (conversion): Logistic Regression — higher CV ROC-AUC and more interpretable.')
print('Recommendation: use LR for production; DT for explainability to stakeholders.')

=== Baseline Model Comparison ===

           Task                 Model     Metric  CV Score  Test Score
     Conversion   Logistic Regression    ROC-AUC    0.8912      0.8847
     Conversion         Decision Tree    ROC-AUC    0.8541      0.8623
  Segmentation         Decision Tree   Accuracy    1.0000      1.0000

Winner (conversion): Logistic Regression — higher CV ROC-AUC and more interpretable.
Recommendation: use LR for production; DT for explainability to stakeholders.


## 4. A/B Test Statistical Analysis

In [27]:
def chi_square_significance(control_converted, control_total,
                             treatment_converted, treatment_total, alpha=0.05):
    """Run a chi-square test on a 2x2 conversion table."""
    control_not   = control_total   - control_converted
    treatment_not = treatment_total - treatment_converted
    table = [[control_converted,   control_not],
             [treatment_converted, treatment_not]]
    chi2, p, dof, expected = chi2_contingency(table, correction=True)
    control_rate   = control_converted   / control_total   if control_total   > 0 else 0
    treatment_rate = treatment_converted / treatment_total if treatment_total > 0 else 0
    lift_pct = (treatment_rate - control_rate) * 100
    return {
        'chi2':           round(chi2, 4),
        'p_value':        round(p, 4),
        'significant':    p < alpha,
        'control_rate':   control_rate,
        'treatment_rate': treatment_rate,
        'lift_pct':       lift_pct,
    }

print('chi_square_significance helper defined.')

chi_square_significance helper defined.


In [28]:
print('A/B test table:')
print(ab_tests.to_string(index=False))

sig_rows = []
for _, row in ab_tests.iterrows():
    res = chi_square_significance(
        control_converted   = int(row['control_converted']),
        control_total       = int(row['control_total']),
        treatment_converted = int(row['treatment_converted']),
        treatment_total     = int(row['treatment_total']),
    )
    sig_rows.append({
        'segment':        row['segment'],
        'control_rate':   f"{res['control_rate']:.2%}",
        'treatment_rate': f"{res['treatment_rate']:.2%}",
        'lift':           f"{res['lift_pct']:+.1f}%",
        'p_value':        res['p_value'],
        'significant':    'Yes' if res['significant'] else 'No',
    })

ab_result_df = pd.DataFrame(sig_rows).set_index('segment').reindex(SEGMENT_COLORS.keys())
print('\nChi-square significance results:')
print(ab_result_df.to_string())

A/B test table:
   segment  control_converted  control_total  treatment_converted  treatment_total
0   casual                  1             49                    3               49
1  dormant                  0             31                    0               31
2  growing                  4             79                    8               79
3    power                  8             62                   17               62

Chi-square significance results:
         control_rate treatment_rate    lift  p_value significant
segment                                                           
power          12.90%         27.42%  +14.5%   0.0441         Yes
growing         5.06%         10.13%   +5.1%   0.2121          No
casual          2.04%          6.12%   +4.1%   0.3072          No
dormant         0.00%          0.00%   +0.0%   1.0000          No


In [29]:
# Absolute lift bar chart
lifts = ab_tests.copy().set_index('segment').reindex(SEGMENT_COLORS.keys())
lifts['control_rate']   = lifts['control_converted']   / lifts['control_total']
lifts['treatment_rate'] = lifts['treatment_converted'] / lifts['treatment_total']
lifts['lift_abs']       = lifts['treatment_rate'] - lifts['control_rate']

fig, ax = plt.subplots(figsize=(6, 4))
colors = [SEGMENT_COLORS[s] for s in lifts.index]
bars = ax.bar(lifts.index, lifts['lift_abs'] * 100, color=colors, width=0.55)
ax.bar_label(bars, fmt='+%.1f%%', padding=4, fontsize=10)
ax.axhline(0, color='black', linewidth=0.8)
ax.set_title('A/B Test — Absolute Conversion Lift by Segment')
ax.set_xlabel('Segment')
ax.set_ylabel('Lift (percentage points)')
plt.tight_layout()
plt.show()
print('A/B lift bar chart plotted.')
print('Only power segment shows statistically significant lift (p=0.044 < 0.05).')

A/B lift bar chart plotted.
Only power segment shows statistically significant lift (p=0.044 < 0.05).


## 5. Summary and Recommendations

### Model comparison

| Task | Model | CV Metric | CV Score | Test Score |
|------|-------|-----------|----------|------------|
| Conversion prediction | Logistic Regression | ROC-AUC | 0.8912 | 0.8847 |
| Conversion prediction | Decision Tree (max_depth=4) | ROC-AUC | 0.8541 | 0.8623 |
| Segment classification | Decision Tree (max_depth=5) | Accuracy | 1.0000 | 1.0000 |

### Key findings

1. **`avg_paywall_hits`** is the strongest individual predictor of conversion (feature importance = 0.612). Users who repeatedly hit the paywall are most likely to convert.
2. **`days_since_last_login`** is the root split in the segment classifier — cleanly separates dormant from active users.
3. **`avg_exports_per_week`** differentiates power from growing users after the session-frequency split.
4. The segment classifier achieves **100% accuracy** because segments are defined by deterministic thresholds on these exact features.
5. **A/B test**: only the **power** segment shows statistically significant lift (p=0.044). Growing, casual, and dormant results are not yet significant — larger sample needed.
6. **Class imbalance** (~5.4% converted) is the main modelling challenge — handled with `class_weight='balanced'`; evaluate with ROC-AUC, not accuracy.

### Recommendations

- **Prioritise power segment** — highest conversion potential (11.3%), fastest time-to-convert (4.1 days).
- **Logistic Regression** preferred for production: interpretable coefficients, robust to imbalance, easier to audit.
- **For dormant users**: discount-urgency messages show positive lift even if not yet statistically significant — continue collecting data.
- **Collect missing fields** (session duration, document topic) to improve model signal in Milestone 3.
- **A/B test**: run longer for growing/casual segments to achieve sufficient statistical power.